src/update-feeds-loops/submit-transactions.ts

```diff
@@ -91,9 +92,109 @@ export const submitUpdate = async (
     gasLimit: gasLimit.toString(),
     nonce,
   });
-  return api3ServerV1
-    .connect(connectedSponsorWallet)
-    .tryMulticall.send(dataFeedUpdateCalldatas, { gasPrice, gasLimit, nonce });
+  return api3ServerV1.connect(sponsorWallet).tryMulticall.send(dataFeedUpdateCalldatas, { gasPrice, gasLimit, nonce });
+};
+
+export const submitBatchTransaction = async (
+  chainId: string,
+  providerName: string,
+  provider: ethers.JsonRpcProvider,
+  api3ServerV1: Api3ServerV1,
+  updatableDataFeeds: UpdatableDataFeed[],
+  blockNumber: number
+) => {
+  const {
+    config: { chains, sponsorWalletMnemonic, walletDerivationScheme },
+  } = getState();
+
+  const decodedDapiNames = updatableDataFeeds.map(({ dataFeedInfo: { decodedDapiName } }) => decodedDapiName);
+  const dataFeedIds = updatableDataFeeds.map(({ dataFeedInfo: { dataFeedId } }) => dataFeedId);
+
+  const { dataFeedUpdateInterval, fallbackGasLimit } = chains[chainId]!;
+  const dataFeedUpdateIntervalMs = dataFeedUpdateInterval * 1000;
+
+  return logger.runWithContext({ dapiNames: decodedDapiNames, dataFeedIds, blockNumber }, async () => {
+    // NOTE: We use go mainly to set a timeout for the whole update process. We expect the function not to throw and
+    // handle errors internally.
+    const goUpdate = await go(
+      async () => {
+        logger.debug('Getting derived sponsor wallet.');
+        const sponsorWallet = getDerivedSponsorWallet(sponsorWalletMnemonic, { walletDerivationScheme }).connect(
+          provider
+        );
+        const sponsorWalletAddress = sponsorWallet.address as Address;
+
+        logger.debug('Getting nonce.');
+        const goNonce = await go(async () => provider.getTransactionCount(sponsorWalletAddress, 'latest'));
+        if (!goNonce.success) {
+          logger.warn(`Failed to get nonce.`, goNonce.error);
+          return null;
+        }
+        const nonce = goNonce.data;
+
+        logger.debug('Getting recommended gas price.');
+        const gasPrice = getRecommendedGasPrice(chainId, providerName, sponsorWalletAddress, dataFeedIds);
+        if (!gasPrice) return null;
+
+        logger.debug('Creating calldatas.');
+        const dataFeedUpdateCalldatas = updatableDataFeeds.flatMap((updatableDataFeed) =>
+          createUpdateFeedCalldatas(api3ServerV1, updatableDataFeed)
+        );
+
+        logger.debug('Estimating batch update gas limit.');
+        const gasLimit = await estimateMulticallGasLimit(api3ServerV1, dataFeedUpdateCalldatas, fallbackGasLimit);
+        if (!gasLimit) return null;
+
+        const goSubmitUpdate = await go(async () => {
+          logger.info('Updating data feeds in batch.', {
+            sponsorWalletAddress,
+            gasPrice: gasPrice.toString(),
+            gasLimit: gasLimit.toString(),
+            nonce,
+          });
+          return api3ServerV1
+            .connect(sponsorWallet)
+            .tryMulticall.send(dataFeedUpdateCalldatas, { gasPrice, gasLimit, nonce });
```

Comment:
Btw. we should only use mutlicall in case there are multiple calldatas. It is possible that only a single beacon will be updated and this should be called directly. Feel free to open separate issue for that.

Reply:
Right, good catch. Last night I was trying to make both `submitTransaction()` and `submitBatchTransaction()` call a single `submitUpdate()` function that handles both cases based on [this comment](https://github.com/api3dao/airseeker-v2/pull/296#discussion_r1593670739) so maybe I can try avoid sending a tryMulticall in there as well